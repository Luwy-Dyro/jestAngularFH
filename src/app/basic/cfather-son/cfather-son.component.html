<h5>Component Soon</h5>


<code>{{client | json}}</code>

<br><br>

<button
    class="btn btn-outline-primary"
    (click)="onChange(10)"
    data-test="btn-changeId"

>
Cambiar ID

</button>

<button
    data-test="btn-delete"
    class="btn btn-outline-danger"
    (click)="onDelete()"
>
Borrar Client

</button>