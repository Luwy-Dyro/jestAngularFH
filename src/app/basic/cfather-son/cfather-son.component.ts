import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Client } from '../interfaces/client';

@Component({
  selector: 'app-cfather-son',
  templateUrl: './cfather-son.component.html',
  styleUrls: ['./cfather-son.component.scss']
})
export class CfatherSonComponent {

  @Input() client?: Client
  @Output() onDeleteClient = new EventEmitter();
  @Output() onUpdatedClient = new EventEmitter<Client>() ;

  onDelete(){
    this.client = undefined;
    this.onDeleteClient.emit()
  }

  onChange(newId: number){

    if(!this.client) return;

    //this.client.id = newId
    this.client = {
      ...this.client,
      id: newId
    }

    this.onUpdatedClient.emit({...this.client})


  }

}
