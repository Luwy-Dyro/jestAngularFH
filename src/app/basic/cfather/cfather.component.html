<h1>Component father</h1>

<button 
    class="btn btn-primary"    
    (click)="onSetClient('Luis')"
>
    Load Client
</button>

<br><br>

<code class="mt-2">
    {{client | json}}
</code>

<hr>

<app-cfather-son [client]="client" 
        (onDeleteClient)="client = undefined" 
        (onUpdatedClient)="client = $event" >
        <!-- //$event, asi se recibe el objeto q emit cuando dispara -->
</app-cfather-son>