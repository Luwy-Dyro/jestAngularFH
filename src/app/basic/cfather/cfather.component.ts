import { Component, Input, OnInit } from '@angular/core';
import { Client } from '../interfaces/client';

@Component({
  selector: 'app-cfather',
  templateUrl: './cfather.component.html',
  styleUrls: ['./cfather.component.scss']
})
export class CfatherComponent implements OnInit{
  
  
  
  public client?: Client


  ngOnInit() {
  
  }

  onSetClient(name: string){

    this.client = {
      id: 1,
      name
    }    

  }


}
