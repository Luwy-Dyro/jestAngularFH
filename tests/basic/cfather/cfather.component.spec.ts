import { ComponentFixture, TestBed } from '@angular/core/testing';

import { CfatherComponent } from '../../../src/app/basic/cfather/cfather.component';
import { CfatherSonComponent } from '../../../src/app/basic/cfather-son/cfather-son.component';
import { By } from '@angular/platform-browser';

describe('CfatherComponent', () => {
  let component: CfatherComponent;
  let fixture: ComponentFixture<CfatherComponent>;
  let compiled: HTMLElement

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ CfatherComponent, CfatherSonComponent ]
    })
    .compileComponents();

    fixture = TestBed.createComponent(CfatherComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

    compiled = fixture.nativeElement
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('debe hacer match con snapchot', () => {
    expect(compiled).toMatchSnapshot();
  });


  it('debe de establecer el cliente con el nombre inidicado', () => {
    
    component.onSetClient('Pepe');
    fixture.autoDetectChanges()

    const codeDiv = compiled.querySelector('.mt-2')
    //console.log(codeDiv?.textContent);
    expect(codeDiv?.textContent).toContain('"name"')
    

  });


  
  it('debe borrar cliente si se emite onDeleteClient', () => {
    
    component.client = {id:1, name: 'Eduardo'};
    fixture.detectChanges();

    const sonDebugElement = fixture.debugElement.query( By.directive(CfatherSonComponent) )

    const sonComponent: CfatherSonComponent = sonDebugElement.componentInstance
    
    //console.log(sonComponent.client);
    sonComponent.onDeleteClient.emit();
    expect( component.client).toBe(undefined)    

    //expect(codeDiv?.textContent).toContain('"name"')
    

  });


  it('debe actualizar cliente si se emite onUpdatedClient', () => {
    
    component.client = {id:1, name: 'Jose'};
    fixture.detectChanges();

    const sonDebugElement = fixture.debugElement.query( By.directive(CfatherSonComponent) )
    const sonComponent: CfatherSonComponent = sonDebugElement.componentInstance
    
    //console.log(sonComponent.client);
    sonComponent.onUpdatedClient.emit({id: 10, name: 'Jose'});
    expect( component.client).toEqual({id: 10, name: 'Jose'})    

    //expect(codeDiv?.textContent).toContain('"name"')
    

  });


});

