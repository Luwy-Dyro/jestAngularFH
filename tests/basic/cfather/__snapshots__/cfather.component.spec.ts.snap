// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CfatherComponent debe hacer match con snapchot 1`] = `
<div
  id="root1"
>
  <h1>
    Component father
  </h1>
  <button
    class="btn btn-primary"
  >
     Load Client

  </button>
  <br />
  <br />
  <code
    class="mt-2"
  >
     

  </code>
  <hr />
  <app-cfather-son>
    <h5>
      Component Soon
    </h5>
    <code>
      
    </code>
    <br />
    <br />
    <button
      class="btn btn-outline-primary"
      data-test="btn-changeId"
    >
      
Cambiar ID 
    </button>
    <button
      class="btn btn-outline-danger"
      data-test="btn-delete"
    >
      
Borrar Client 
    </button>
  </app-cfather-son>
</div>
`;
